packages/@tailwindcss-upgrade/src/utils/packages.ts

```diff
@@ -1,25 +1,127 @@
+  }
+}
+
+class BunPackageManager extends PackageManager {
+  add(packages: string[], location: 'dependencies' | 'devDependencies') {
+    let args = packages.slice()
+
+    if (location === 'devDependencies') {
+      args.unshift('--development')
+    }
 
-export async function pkg(command: string, base: string): Promise<Buffer | void> {
-  let packageManager = await detectPackageManager(base)
-  if (!packageManager) {
-    if (!didWarnAboutPackageManager) {
-      didWarnAboutPackageManager = true
-      warn('Could not detect a package manager. Please manually update `tailwindcss` to v4.')
+    return this.exec(`bun add ${args.join(' ')}`)
+  }
+
+  remove(packages: string[]) {
+    return this.exec(`bun remove ${packages.join(' ')}`)
+  }
+}
+
+class YarnPackageManager extends PackageManager {
+  add(packages: string[], location: 'dependencies' | 'devDependencies') {
+    let args = packages.slice()
+
+    if (location === 'devDependencies') {
+      args.unshift('--dev')
     }
-    return
+
+    return this.exec(`yarn add ${args.join(' ')}`)
+  }
+
+  remove(packages: string[]) {
+    return this.exec(`yarn remove ${packages.join(' ')}`)
   }
-  return execSync(`${packageManager} ${command}`, {
-    cwd: base,
+}
+
+class PnpmPackageManager extends PackageManager {
+  add(packages: string[], location: 'dependencies' | 'devDependencies') {
+    let args = packages.slice()
+
+    if (location === 'devDependencies') {
+      args.unshift('--save-dev')
+    }
+
+    return this.exec(`pnpm add ${args.join(' ')}`)
+  }
+
+  remove(packages: string[]) {
+    return this.exec(`pnpm remove ${packages.join(' ')}`)
+  }
+}
+
+class NpmPackageManager extends PackageManager {
+  add(packages: string[], location: 'dependencies' | 'devDependencies') {
+    let args = packages.slice()
+
+    if (location === 'devDependencies') {
+      args.unshift('--save-dev')
+    }
+
+    return this.exec(`npm install ${args.join(' ')}`)
+  }
+
+  remove(packages: string[]) {
+    return this.exec(`npm remove ${packages.join(' ')}`)
+  }
+}
+
+let packageManagers = new DefaultMap((base) => {
+  return new DefaultMap<string, PackageManager>((pm) => {
+    switch (pm) {
+      case 'bun':
+        return new BunPackageManager(base)
+      case 'yarn':
+        return new YarnPackageManager(base)
+      case 'pnpm':
+        return new PnpmPackageManager(base)
+      case 'npm':
+        return new NpmPackageManager(base)
+      default:
+        throw new Error(`Unknown package manager: ${pm}`)
+    }
   })
+})
+
+export function pkg(base: string) {
+  return {
+    async add(packages: string[], location: 'dependencies' | 'devDependencies' = 'dependencies') {
+      let packageManager = await detectPackageManager(base)
+      return packageManager.add(packages, location)
+    },
+    async remove(packages: string[]) {
+      let packageManager = await detectPackageManager(base)
+      return packageManager.remove(packages)
+    },
```

Comment:
We could return the packageManager directly, but then the usage API would look like this:
```diff
- pkg(base).add(['@tailwindcss/postcss@next'])
+ pkg(base).then(pm => pm.add(['@tailwindcss/postcss@next']))
```

Which isn't as nice to work with.